bindinfo/handle.go

```diff
@@ -838,57 +837,111 @@ func GenerateBindSQL(ctx context.Context, stmtNode ast.StmtNode, planHint string
-	// So here we would remove the hint and inject the current best plan hint.
-	hint.BindHint(stmtNode, &hint.HintsSet{})
-	bindSQL := utilparser.RestoreWithDefaultDB(stmtNode, defaultDB, "")
-	if bindSQL == "" {
-		return ""
-	}
+	// We can place hint at update, delete, select stmt block,there are three combinations follows:
+	// updateStmt-selectStmt,deleteStmt-selectStmt,insertStmt-selectStmt,
+	// so when these are present，we need a way to prevent the hint of wrong placement,
+	// eg `update /*+ use_index(@upd_1 t1 idx_b) t1 set b = 1 where  (a in (select /*+use_index(@sel_1 t2  idx_ba)*/ a from t2 where b = 1) )`
+	// if we only order hint by parse offset from qb_name then @upd_1 and @sel_1 will be same offset,
+	// it's incorrect,so we need a way to deal this，so introduce the startBlockOffset，
+	// the startBlockOffset is the start of the offset to place hint.now if we get @upd_1 and @sel_1,we will set startBlockOffset=0,
+	// because ParseOffsetFromBlockName only parse qbName which Prefix with  'sel_',
+	// so upd_1 will be parsed to startBlockOffset that is 0,the @sel_1 will be parse to 1,it right result for us
+	var blockStartOffset int
+	var afterProcessFunc func(string) string
 	switch n := stmtNode.(type) {
 	case *ast.DeleteStmt:
-		deleteIdx := strings.Index(bindSQL, "DELETE")
-		// Remove possible `explain` prefix.
-		bindSQL = bindSQL[deleteIdx:]
-		return strings.Replace(bindSQL, "DELETE", fmt.Sprintf("DELETE /*+ %s*/", planHint), 1)
+		blockStartOffset = 0
+		afterProcessFunc = func(bindSQL string) string {
+			deleteIdx := strings.Index(bindSQL, "DELETE")
+			// Remove possible `explain` prefix.
+			return bindSQL[deleteIdx:]
+		}
 	case *ast.UpdateStmt:
-		updateIdx := strings.Index(bindSQL, "UPDATE")
-		// Remove possible `explain` prefix.
-		bindSQL = bindSQL[updateIdx:]
-		return strings.Replace(bindSQL, "UPDATE", fmt.Sprintf("UPDATE /*+ %s*/", planHint), 1)
+		blockStartOffset = 0
+		afterProcessFunc = func(bindSQL string) string {
+			updateIdx := strings.Index(bindSQL, "UPDATE")
+			// Remove possible `explain` prefix.
+			return bindSQL[updateIdx:]
+		}
 	case *ast.SelectStmt:
-		var selectIdx int
-		if n.With != nil {
-			var withSb strings.Builder
-			withIdx := strings.Index(bindSQL, "WITH")
-			restoreCtx := format.NewRestoreCtx(format.RestoreStringSingleQuotes|format.RestoreSpacesAroundBinaryOperation|format.RestoreStringWithoutCharset|format.RestoreNameBackQuotes, &withSb)
-			restoreCtx.DefaultDB = defaultDB
-			err := n.With.Restore(restoreCtx)
-			if err != nil {
-				logutil.BgLogger().Debug("[sql-bind] restore SQL failed", zap.Error(err))
-				return ""
+		blockStartOffset = 1
+		afterProcessFunc = func(bindSQL string) string {
+			var selectIdx int
+			if n.With != nil {
+				selectIdx = strings.Index(bindSQL, "WITH")
+			} else {
+				selectIdx = strings.Index(bindSQL, "SELECT")
+			}
+			// Remove possible `explain` prefix.
+			return bindSQL[selectIdx:]
+		}
+	case *ast.SetOprStmt:
```

Comment:
Please add more tests about this without `CTE`.

Reply:
done